tensorboard/webapp/widgets/line_chart_v2/lib/formatter_test.ts

```diff
@@ -0,0 +1,143 @@
+        expect(numberFormatter.formatLong(1)).toBe('1');
+        expect(numberFormatter.formatLong(5)).toBe('5');
+        expect(numberFormatter.formatLong(-100.4)).toBe('-100.4');
+        expect(numberFormatter.formatLong(3.01)).toBe('3.01');
+        expect(numberFormatter.formatLong(9999)).toBe('9,999');
+        expect(numberFormatter.formatLong(0.09)).toBe('0.09');
+        expect(numberFormatter.formatLong(1.004e6)).toBe('1,004,000');
+        expect(numberFormatter.formatLong(-1.004e6)).toBe('-1,004,000');
+        expect(numberFormatter.formatLong(0.00005)).toBe('0.00005');
+        expect(numberFormatter.formatLong(1e5 + 0.00005)).toBe('100,000.00005');
+      });
+    });
+  });
+
+  describe('#relativeTimeFormatter', () => {
+    for (const {name, fn} of [
+      {name: 'formatTick', fn: relativeTimeFormatter.formatTick},
+      {name: 'formatShort', fn: relativeTimeFormatter.formatShort},
+      {name: 'formatLong', fn: relativeTimeFormatter.formatLong},
+    ]) {
+      describe(name, () => {
+        it('formats time difference in appropriate unit', () => {
+          expect(fn(0)).toBe('0');
+          expect(fn(100)).toBe('100ms');
+          expect(fn(999)).toBe('999ms');
+          expect(fn(1000)).toBe('1sec');
+          expect(fn(4023)).toBe('4.023sec');
+          expect(fn(60023)).toBe('1min');
+          expect(fn(61523)).toBe('1.025min');
+          expect(fn(3700000)).toBe('1.028hr');
+          expect(fn(86400000 * 3)).toBe('3day');
+          expect(fn(31536000000 * 5)).toBe('5yr');
+        });
+
+        it('formats negative time difference in appropriate unit', () => {
+          expect(fn(-100)).toBe('-100ms');
+          expect(fn(-999)).toBe('-999ms');
+          expect(fn(-1000)).toBe('-1sec');
+          expect(fn(-4023)).toBe('-4.023sec');
+          expect(fn(-60023)).toBe('-1min');
+          expect(fn(-61523)).toBe('-1.025min');
+          expect(fn(-3700000)).toBe('-1.028hr');
+          expect(fn(-86400000 * 3)).toBe('-3day');
+          expect(fn(-31536000000 * 5)).toBe('-5yr');
+        });
+      });
+    }
+  });
+
+  describe('#timeFormatter', () => {
+    beforeEach(() => {
+      TEST_ONLY.setLocale('en-US');
+    });
+
+    describe('#formatTick', () => {
+      it('returns tick format using d3', () => {
+        expect(
+          timeFormatter.formatTick(new Date('2020-1-5 13:23').getTime())
+        ).toBe('01:23');
+        expect(
+          timeFormatter.formatTick(new Date('2020-1-5 5:00').getTime())
+        ).toBe('05 AM');
+        expect(timeFormatter.formatTick(new Date('2020-1-5').getTime())).toBe(
+          'Jan 05'
+        );
+        expect(timeFormatter.formatTick(new Date('2019-1-1').getTime())).toBe(
+          '2019'
+        );
+      });
+    });
+
+    describe('formatShort', () => {
+      it('formats using localization', () => {
+        expect(
+          timeFormatter.formatShort(new Date('2020-1-5 13:23').getTime())
+        ).toBe('1/5/20, 1:23:00 PM');
```

Comment:
Can we get this to spit out unambiguous dates? `1/5/20` is a good
example of the problem. I’d prefer `2020-01-05` myself, but also
amenable to `Jan 5, 2020` or similar. <https://xkcd.com/1179/>

Even if this is meant to use “the user’s current locale”, I personally
don’t find that very compelling. If you and I both use TensorBoard.dev,
and I load the plot and it says `1/5/20`, and I take a screenshot and
send it to you, and when you open the plot it says `5/1/20`, that’s
confusing.